### ISSUE TITLE
Create view with INSERT only privilege behaves different with MySQL 8.0

### ISSUE THREAD

**Issue by erjiaqing** [pingcap/tidb#9796]
## Bug Report

Please answer these questions before submitting your issue. Thanks!

1. What did you do?
If possible, provide a recipe for reproducing the error.

root:
```sql
create database test01;
create database test02;
create user test_u1@'%';
create table test01.t1 (c1 int);
grant insert on test01.t1 to test_u1@'%';
grant all privileges test02.* to test_u1@'%';
```
test_u1:
```sql
use test02;
create view v2 as select * from test01.t1;
```

2. What did you expect to see?

No error. MySQL throws error when execute `select * from v2`.

3. What did you see instead?

`ERROR 1142 (42000): SELECT command denied to user '127.0.0.1'@'test_u1' for table 't1'` when `create view`

4. What version of TiDB are you using (`tidb-server -V` or run `select tidb_version();` on TiDB)?

`v3.0.0-beta-231-g20463d6da`

**Comment by erjiaqing** [comment]
I think the expected behavior is throw the error, and in MySQL document:

> The CREATE VIEW statement requires the CREATE VIEW privilege for the view, and some privilege for each column selected by the SELECT statement. For columns used elsewhere in the SELECT statement, you must have the SELECT privilege. If the OR REPLACE clause is present, you must also have the DROP privilege for the view. CREATE VIEW might also require the SET_USER_ID or SUPER privilege, depending on the DEFINER value, as described later in this section.

(via https://dev.mysql.com/doc/refman/8.0/en/create-view.html)

But in a clean install of MySQL 8.0:

![](https://user-images.githubusercontent.com/5517838/54590622-78c67300-4a63-11e9-9c48-adf3c34eca0b.png)

I believe this should be an bug of MySQL, so should we follow the bug or not?

@morgo @kolbe PTAL.

**Comment by erjiaqing** [comment]
I noticed that the document used "some privilege", does it mean that having "SELECT" or some other privileges on that db/table/column is enough for creating the view?

**Comment by morgo** [comment]
@erjiaqing I agree with you, the behavior is not compatible with MySQL.  It took me some time to understand the test-case, I re-wrote it here while wrapping my head around it:

```
# mysql -uroot
drop database if exists d1;
drop database if exists d2;
drop user if exists u1;
create database d1;
create database d2;
create user u1;
create table d1.t1 (c1 int);
grant insert on d1.t1 to u1;
grant all privileges on d2.* to u1;

# mysql -uu1 d2
create view v1 as select * from d1.t1;
```

As you said, the `create view` step passes in MySQL, since it can not be assumed that you want to SELECT from a view.  You may only want to insert into it (which there are permissions for):

```
mysql [localhost:5725] {u1} (d2) > INSERT INTO v1 VALUES (1);
Query OK, 1 row affected (0.00 sec)

mysql [localhost:5725] {u1} (d2) > select * from v1;
ERROR 1356 (HY000): View 'd2.v1' references invalid table(s) or column(s) or function(s) or definer/invoker of view lack rights to use them
```

Where it gets naunced of course, is that TiDB does not have insertable views (they are only readable), so there is no other use other than `SELECT`!

```
mysql> INSERT INTO v1 VALUES (1);
ERROR 1105 (HY000): insert into view v1 is not supported now.
```
I still think we should fix it, but it is probably not high priority because of this nuance.

**Comment by morgo** [comment]
I put some more thought in this, and it may actually block `mysqldump`/`mydumper` from restoring.  There could be a `DEFINER` set to a user which only has insert privileges.  Even if the view can currently not effectively be used, we should make sure mysqldump can restore.

**Comment by AndrewDi** [comment]
@morgo I think mysql's behavior is wired, when user didn't have privilge on a base table, it can create a view on it...

**Comment by morgo** [comment]
> @morgo @erjiaqing I think mysql's behavior is wired, when user didn't have privilge on a base table, it can create a view on it...

The user must have a privilege, but that privilege does not need to be `select`. It could be `insert` instead.
